connlib: Improve FFI bridges for Apple and Android

francesca64 opened the pull request:
This makes it possible to build the Apple/Android FFI bridges and integrate them with their respective client apps.

francesca64 committed:
connlib: Fix macOS compilation errors

francesca64 committed:
connlib: Remove `SwiftEnum` macro

`swift-bridge` works by parsing Rust source code from the filesystem,
which happens prior to macro expansion. Therefore, using macros for
this is more trouble than it's worth.

francesca64 committed:
connlib: Add `&self` to `Callbacks` associated fns

francesca64 committed:
connlib: Fix Android compilation errors

francesca64 committed:
connlib: Modernize Android bridge code

francesca64 committed:
connlib: Fix Apple framework build

francesca64 committed:
connlib: Add script for local Apple development

francesca64 committed:
connlib: Fix macOS min version mismatch

francesca64 committed:
connlib: Fix `build-rust.sh` unused linker argument error

francesca64 committed:
connlib: Add logging to Apple `onError`

francesca64 committed:
connlib: Add helpful context to Apple local dev script

thomaseizinger commented on rust/connlib/libs/tunnel/src/tun_android.rs:
```diff
@@ -1,20 +1,87 @@
+        // TODO: This won't actually work for non-root users...
+        let fd = unsafe { open(b"/dev/net/tun\0".as_ptr() as _, O_RDWR) };
+        // TODO: everything!
+        if fd == -1 {
+            Err(Error::Io(std::io::Error::last_os_error()))
+        } else {
+            Ok(Self { fd })
+        }
```
It is 100% preference so feel free to discard but I find that using early return and avoiding else to be much easier to read:

```suggestion
        if fd == -1 {
            return Err(Error::Io(std::io::Error::last_os_error()))
        }
        
        Ok(Self { fd })
```

francesca64 replied:
I empathize with your viewpoint, but I consider early return to be a larger hammer, and branching has the additional benefit of making it easier to visualize potential combinator golfs/etc.

thomaseizinger commented on rust/connlib/libs/common/src/session.rs:
```diff
@@ -31,9 +31,9 @@ pub trait ControlSession<T, U> {
 ///
 /// A session is created using [Session::connect], then to stop a session we use [Session::disconnect].
-pub struct Session<T, U, V, R, M> {
+pub struct Session<T, U, V, R, M, CB: Callbacks + 'static> {
```
Having trait bounds on structs can become really cumbersome because they are quite infectious to the rest of the codebase. Unless you are referencing associated types (`Callbacks` doesn't have any from what I can see?) somewhere, I wouldn't do it.

Traits are about behaviour and behaviour is specified on `impl` blocks. `struct` declarations should be for layout only.

It is weird initially to have a type-parameter to be "too open" for what you want users to use it but in reality, they can't do anything with a "wrongly" instantiated type anyway if you have the correct bounds on your `impl` blocks which is where you need them anyway to actually implement your stuff and call the trait functions.

francesca64 replied:
I understand your point, but I don't see the merit in `Session` having more degrees of freedom than we actually want or use. The type definition reads loud and clear: a `Session` where `CB` does not implement `Callbacks` is not a `Session`.

The fact that Rust requires trait bounds to be repeated does not change the fact that this type definition is my explicit intention. While your proposal may reduce visual noise, it increases conceptual/mental-model noise by introducing a degenerate case for `Session` types.

thomaseizinger replied:
> I understand your point, but I don't see the merit in `Session` having more degrees of freedom than we actually want or use.

Yeah, whether or not repeating the trait bounds is annoying depends a lot on how many layers of nesting you've got.

If your trait is object-safe, another way would be to use a trait-object instead of a type-parameter. That is equally expressive from a readability PoV but results in simpler type definitions.

conectado replied:
I'd rather not introduce trait-objects if we don't need to, that  would indicate some kind of runtime polymorphism that we don't actually need.

conectado replied:
> The fact that Rust requires trait bounds to be repeated does not change the fact that this type definition is my explicit intention. While your proposal may reduce visual noise, it increases conceptual/mental-model noise by introducing a degenerate case for Session types.

I agree with this but then we might want to remove the `'static` bound? That's just an implementation detail that we might even want to change in the future.

thomaseizinger commented on rust/connlib/libs/common/src/session.rs:
```diff
@@ -51,38 +51,39 @@ pub struct TunnelAddresses {
+pub trait Callbacks: Clone + Send + Sync {
     /// Called when there's a change in the resource list.
-    fn on_update_resources(resource_list: ResourceList);
+    fn on_update_resources(&self, resource_list: ResourceList);
```
This forces interior-mutablity on all implementors. It might be more ergonomic to either have connlib have a single owner of this or make the trait `&mut self` and `Arc<Mutex>` it inside connlib.

francesca64 replied:
We decided that the burden of thread-safety/etc. is on the non-Rust side: https://github.com/firezone/product/blob/main/architecture/madr.md#2023_06_14-client-app-call

On the Rust side, the underlying Apple/Android callback handler is just an opaque pointer to some Swift/Kotlin object.

thomaseizinger commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -71,8 +70,8 @@ pub unsafe extern "system" fn Java_dev_firezone_connlib_Session_connect(
-        Ok(res) => trace!("onSetTunnelAddresses returned {:?}", res),
-        Err(e) => error!("Failed to call setTunnelAddresses: {:?}", e),
+        Ok(res) => log::trace!("`onSetTunnelAddresses` returned `{res:?}`"),
+        Err(err) => log::error!("Failed to call `setTunnelAddresses`: {err}"),
```
+1 for referencing `trace!` etc with `log::`. I much prefer that too because it avoids import lines and actually reads quite nice.

thomaseizinger commented on rust/connlib/clients/apple/Sources/Connlib/CallbackHandler.swift:
```diff
@@ -97,5 +97,7 @@ public class CallbackHandler {
 
     func onError(error: SwiftConnlibError, error_type: SwiftErrorType) {
         // TODO: handle/report errors
```
Resolved now?

francesca64 replied:
Errors need to be propagated to the client; the logging is only helpful for testing.

thomaseizinger reviewed (dismissed):
Super clean commit separation, made it very easy to review this!

I can't comment on any of the ffi stuff :)

francesca64 committed:
ci: Don't check/test connlib Apple FFI bridge on non-Apple platforms

jamilbk commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -1,43 +1,44 @@
     }
 
-    fn on_set_tunnel_adresses(_tunnel_addresses: TunnelAddresses) {
+    fn on_set_tunnel_adresses(&self, _tunnel_addresses: TunnelAddresses) {
```
We [may have renamed](https://github.com/firezone/product/blob/main/architecture/madr.md#2023_06_19-grooming) this callback to `onConnect` in the last grooming. Will defer to you though -- I don't feel strongly Re: naming conventions

francesca64 replied:
Absolutely; I was trying to avoid adding something else onto an already large-ish PR, but that's a small enough addition that it'd be worth me adding now just to avoid additional churn for the clients.

jamilbk replied:
Ah gotcha!

jamilbk reviewed (commented):
Just the note regarding `onConnect`. Otherwise LGTM, looks a lot cleaner from what I can tell!

francesca64 committed:
ci: Fix relay test step

francesca64 committed:
connlib: `on_set_tunnel_addresses` -> `on_connect`

francesca64 committed:
rust: Update patched `ring` rev

The upstream `cc` fork force-pushed, so the old rev is gone.

francesca64 committed:
ci: Install build tools for *ring* on Windows

francesca64 committed:
ci: Fix *ring* build tool path

conectado reviewed (commented):
looks great! will approve once the CI is passing.

roop commented on rust/connlib/clients/apple/build-xcframework-dev.sh:
```diff
@@ -0,0 +1,27 @@
+
+echo $SRC_ROOT
+
+for sdk in macosx; do
```
Maybe we should make this build for both macOS and iOS by default?
```suggestion
for sdk in macosx iphoneos; do
```

jamilbk replied:
Ah, thanks for catching this @roop; yes on Apple I think we decided to develop for both `iphoneos` and `macosx` targets in parallel due to how much overlap there is. The main difference I believe is only the UI (and the hurdles involved with debugging on the iphonesimulator)

roop commented on rust/connlib/clients/apple/build-xcframework-dev.sh:
```diff
@@ -0,0 +1,27 @@
+
+rm -rf ./Connlib.xcframework
+xcodebuild -create-xcframework \
+  -framework ./connlib-macosx.xcarchive/Products/Library/Frameworks/connlib.framework \
```
```suggestion
  -framework ./connlib-iphoneos.xcarchive/Products/Library/Frameworks/connlib.framework \
  -framework ./connlib-macosx.xcarchive/Products/Library/Frameworks/connlib.framework \
```

roop commented on rust/connlib/clients/apple/build-xcframework-dev.sh:
```diff
@@ -0,0 +1,27 @@
+  -output ./Connlib.xcframework
+
+echo "Build successful. Removing temporary archives"
+rm -rf ./connlib-macosx.xcarchive
```
```suggestion
rm -rf ./connlib-iphoneos.xcarchive
rm -rf ./connlib-macosx.xcarchive
```

francesca64 committed:
rust: Update patched `ring` rev again

francesca64 committed:
connlib: Add iOS to local dev script

Co-authored-by: Roopesh Chander <roop@roopc.net>
Signed-off-by: Francesca Lovebloom <franlovebloom@gmail.com>

francesca64 committed:
connlib: Remove `'static` bound from `Session` + `ControlSession`

francesca64 committed:
connlib: Fix Windows warning

francesca64 committed:
connlib: Fix redundant clone in Apple build script

francesca64 committed:
ci: Fix env var usage on Windows

This would need to be `$env:CARGO_ARGS` on Windows, while still being
`$CARGO_ARGS` elsewhere... handling that is too much complexity for how
little repetition this env var saves us.

francesca64 committed:
connlib: Fix outdated Android log message

francesca64 committed:
connlib: Stub out `on_disconnect`

conectado reviewed (dismissed):
CI is passing, woo!

Excited to get this merged

francesca64 committed:
ci: Fix Rust cache workspace

francesca64 committed:
ci: Fix release drafter `commitish`

https://github.com/release-drafter/release-drafter/issues/1125#issuecomment-1471142225

francesca64 committed:
ci: Fix connlib Apple artifact path

francesca64 committed:
ci: codespell ignore `commitish`

conectado reviewed (approved):
:tada:
